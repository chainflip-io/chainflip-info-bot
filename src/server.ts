import { createBullBoard } from '@bull-board/api';
import { BullMQAdapter } from '@bull-board/api/bullMQAdapter.js';
import { FastifyAdapter } from '@bull-board/fastify';
import { type Queue } from 'bullmq';
import fastify from 'fastify';
import { GraphQLClient } from 'graphql-request';
import env from './env.js';
import logger from './utils/logger.js';

const app = fastify({ loggerInstance: logger, disableRequestLogging: true });

export const explorerClient = new GraphQLClient(env.EXPLORER_GATEWAY_URL);
export const lpClient = new GraphQLClient(env.LP_GATEWAY_URL);

export const createServer = (queues: Queue[]) => {
  const serverAdapter = new FastifyAdapter();

  createBullBoard({
    queues: queues.map((q) => new BullMQAdapter(q)),
    serverAdapter,
  });

  const basePath = '/admin/queues';
  serverAdapter.setBasePath(basePath);
  app.register(serverAdapter.registerPlugin(), { prefix: basePath, basePath: '/' });

  app.get('/health', () => ({ status: 'ok' }));

  app.get('/', (req, res) => res.redirect('/admin/queues', 302));

  return app;
};
