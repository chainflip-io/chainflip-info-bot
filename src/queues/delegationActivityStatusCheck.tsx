import { abbreviate } from '@chainflip/utils/string';
import BigNumber from 'bignumber.js';
import { renderToStaticMarkup } from 'react-dom/server';
import { type JobConfig, type DispatchJobArgs, type JobProcessor } from './initialize.js';
import { Bold, ExplorerLink, Line } from '../channels/formatting.js';
import { TokenAmount, UsdValue } from '../channels/formatting.js';
import { platforms } from '../config.js';
import { DelegationActivityType } from '../graphql/generated/explorer/graphql.js';
import { getDelegationActivityByIdQuery } from '../queries/explorer.js';
import { explorerClient } from '../server.js';
import baseLogger from '../utils/logger.js';

const name = 'delegationActivityStatusCheck';
type Name = typeof name;

const logger = baseLogger.child({ queue: name });

type Data = {
  delegationActivityId: number;
};

declare global {
  interface JobData {
    [name]: Data;
  }
}

type DelegationActivity = {
  id: number;
  type: DelegationActivityType;
  txHash?: string | null;
  amount: string;
  valueUsd: string;
  operatorId: number;
  operatorAlias?: string | null;
  operatorIdSs58: string;
};

const buildMessageData = ({
  delegationActivityInfo,
}: {
  delegationActivityInfo: DelegationActivity;
}): Extract<DispatchJobArgs, { name: 'messageRouter' }>[] => {
  const { id, operatorAlias, operatorIdSs58, txHash, amount, valueUsd } = delegationActivityInfo;

  return platforms.map((platform) => ({
    name: 'messageRouter' as const,
    data: {
      platform,
      message: renderToStaticMarkup(
        <>
          <Line>
            New delegation detected: <Bold platform={platform}>#{id}</Bold>
          </Line>
          <Line>
            üë∑‚Äç‚ôÇÔ∏è Operator:{' '}
            <ExplorerLink platform={platform} path={`/operators/${operatorIdSs58}`} prefer="text">
              {operatorAlias ?? abbreviate(operatorIdSs58)}
            </ExplorerLink>
          </Line>
          <Line>
            üì• Delegated Amount:{' '}
            <Bold platform={platform}>
              <TokenAmount amount={BigNumber(amount)} asset="Flip" />
            </Bold>
            <UsdValue amount={BigNumber(valueUsd)} />
          </Line>
          {txHash && (
            <>
              <Line>
                üßæ Transaction refs:{' '}
                <ExplorerLink path={txHash} chain="Ethereum" platform={platform} prefer="text">
                  {txHash}
                </ExplorerLink>
              </Line>
            </>
          )}
        </>,
      ).trimEnd(),
      filterData: { name: 'NEW_DELEGATION' },
    },
  }));
};

const processJob: JobProcessor<Name> = (dispatchJobs) => async (job) => {
  logger.info(`Checking new delegation activity request #${job.data.delegationActivityId}`);

  const delegationActivitiesQuery = await explorerClient.request(getDelegationActivityByIdQuery, {
    id: job.data.delegationActivityId,
  });

  console.log('delegationActivitiesQuery', delegationActivitiesQuery);

  const delegationActivity = delegationActivitiesQuery.allDelegationActivities?.nodes[0];

  if (!delegationActivity) {
    throw new Error(`Delegation activity not found for id ${job.data.delegationActivityId}`);
  }

  const {
    id,
    type,
    txHash,
    amount,
    valueUsd,
    operatorByOperatorId: {
      id: operatorId,
      accountByAccountId: { alias: operatorAlias, idSs58: operatorIdSs58 },
    },
  } = delegationActivity;

  const delegationActivityInfo = {
    id,
    type,
    txHash,
    amount,
    valueUsd,
    operatorId,
    operatorAlias,
    operatorIdSs58,
  };

  await dispatchJobs(buildMessageData({ delegationActivityInfo }));
};

export const config: JobConfig<typeof name> = {
  name,
  processJob,
};
