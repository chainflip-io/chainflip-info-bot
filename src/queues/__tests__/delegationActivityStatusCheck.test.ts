import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';
import { explorerClient } from '../../server.js';
import { config } from '../delegationActivityStatusCheck.js';

const mockGetDelegationActivityByIdQuery = (delegationActivityId: number) => ({
  allDelegationActivities: {
    nodes: [
      {
        id: delegationActivityId,
        type: 'DELEGATE',
        amount: '0',
        valueUsd: '0.000000000000000000000000000000',
        operatorByOperatorId: {
          id: 1,
          accountByAccountId: {
            alias: null,
            idSs58: 'cFMjXCTxTHVkSqbKzeVwJ25TJxLqc1Vn9usPgUGmZhsyvHRQZ',
          },
        },
        txHash: null,
        event: {
          indexInBlock: 21,
          blockId: 436,
          block: {
            timestamp: '2025-09-05T12:43:36+00:00',
          },
        },
      },
    ],
  },
});

describe('delegationActivityStatusCheck', () => {
  beforeEach(() => {
    vi.useFakeTimers();
  });

  afterEach(() => {
    vi.useRealTimers();
  });

  it('check fresh deelgation activity status and send delegation info message', async () => {
    vi.mocked(explorerClient.request).mockResolvedValue(mockGetDelegationActivityByIdQuery(2));

    vi.setSystemTime(new Date('2024-10-25T12:42:30+00:00'));

    const dispatchJobs = vi.fn();

    await config.processJob(dispatchJobs)({
      data: {
        delegationActivityId: 20,
      } as JobData['delegationActivityStatusCheck'],
    } as any);

    expect(dispatchJobs.mock.lastCall).toMatchSnapshot();
  });
});
