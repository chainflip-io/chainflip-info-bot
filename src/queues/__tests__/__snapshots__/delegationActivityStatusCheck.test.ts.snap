// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`delegationActivityStatusCheck > check fresh deelgation activity status and send delegation info message 1`] = `
[
  [
    {
      "data": {
        "filterData": {
          "name": "NEW_DELEGATION",
        },
        "message": "New delegation detected: <strong>#2</strong>
👷‍♂️ Operator: <a href="https://scan.chainflip.io/operators/cFMjXCTxTHVkSqbKzeVwJ25TJxLqc1Vn9usPgUGmZhsyvHRQZ">cFMj…HRQZ</a>
📥 Delegated Amount: <strong>0 FLIP on Ethereum</strong> ($0.00)",
        "platform": "telegram",
      },
      "name": "messageRouter",
    },
    {
      "data": {
        "filterData": {
          "name": "NEW_DELEGATION",
        },
        "message": "New delegation detected: **#2**
👷‍♂️ Operator: [cFMj…HRQZ](https://scan.chainflip.io/operators/cFMjXCTxTHVkSqbKzeVwJ25TJxLqc1Vn9usPgUGmZhsyvHRQZ)
📥 Delegated Amount: **0 FLIP on Ethereum** ($0.00)",
        "platform": "discord",
      },
      "name": "messageRouter",
    },
    {
      "data": {
        "filterData": {
          "name": "NEW_DELEGATION",
        },
        "message": "New delegation detected: #2
👷‍♂️ Operator: cFMj…HRQZ
📥 Delegated Amount: 0 FLIP on Ethereum ($0.00)",
        "platform": "twitter",
      },
      "name": "messageRouter",
    },
  ],
]
`;
